source/language/openpulse.rst

```diff
@@ -0,0 +1,609 @@
+and hardware. At the core of the OpenPulse grammar are the concepts of ``frames``.
+``frames`` are used to schedule the playing of a ``waveform`` or capturing of data
+via a ``channel`` on the target device. A channel is a software abstraction which allows the
+programmer to be agnostic to complexities of the device's underlying pulse generation hardware. It
+is the responsibility of the target device's compiler to map ``frames`` to the applied channels in
+target hardware.
+
+
+Channels
+--------
+
+Channels map to a hardware resource, which can play pulses to manipulate a qubit
+or capture data from the qubit to perform a measurement. There is a many-to-many
+relationship between qubits and channels. One qubit may have multiple channels
+connecting to it. Pulses on different channels would have different physical
+interactions with that qubit. A channel may also have many qubits. For instance,
+a channel could manipulate the coupling between two neighboring qubits, or
+could even reference multiple qubits in a chain.
+
+Channels are vendor-specific and device-specific. It is expected that vendors
+of quantum hardware provide the appropriate channel names and qubit mappings
+as configuration information to end users. Each channel may also have associated
+static settings, such as local-oscillator frequencies, which do not vary
+throughout program execution. Again it is expected that vendors of quantum
+hardware provide a method for manipulating those static settings if appropriate.
+
+There are two kinds of channels: transmit channels (sending input to a quantum
+device) and receive channels (reading output from a quantum device).
+
+A channel is only used to specify the physical resource on which to play a pulse or from which
+to capture data. This specification should be done by providing a mapping between a qubit list +
+name and the configured hardware channel. The hardware can then be accessed as
+OpenPulse ``txchannel``/``rxchannel``'s via "get" functions.
+
+.. code-block:: c
+
+    txch(qubit q0, ..., qubit qn, str name) -> txchannel  // get transmit channel
+    rxch(qubit q0, ..., qubit qn, str name) -> rxchannel  // get receive channel
+
+The qubits must be **physical** qubits. Furthermore, ordering of qubits is important. For instance,
+``txch($0, $1, "control")`` and ``txch($1, $0, "control")`` may be used to implement distinct
+cross-resonance gates. It is also possible to access a channel by its full name, without supplying
+any qubits, if that has between implemented by the vendor. For instance, ``txch("<channel_name>")``
+may refer to a transmit channel with an arbitrary name.
+
+.. code-block:: c
+
+    txchannel d0 = txch($0, "drive");  // channel for driving at qubit $0's freq
+    txchannel cr1_2 = txch($1, $2, "coupler");  // channel for a coupler between two qubits
+    txchannel m2 = txch($2, "measure");  // channel for transmitting measurement stimulus
```

Comment:
Add name-only example of channel declaration...

Reply:
Fixed in da29859.